### ISSUE TITLE
[EPIC][NEUROGOTCHI] Prolog-governed adaptive tool-chain learner + cross-platform companion UI

### ISSUE THREAD

**Issue by lost-rob0t** [lost-rob0t/zara#979]
## Goal

Add a **Pwnagotchi-inspired local companion/learning loop** to Zara without adding Wi-Fi attack behavior or a second agent authority.

The useful part of the inspiration is the visible stateful companion, bounded local learning, mood/status feedback, and incremental adaptation. This issue explicitly does **not** add deauthentication, handshake capture, credential collection, wireless exploitation, or autonomous network attacks.

The feature learns which **combinations of user-input features + recent canonical tool calls** tend to work, then proposes the next tool or short tool-chain candidate. **Prolog remains the policy/decision gate and Zara's existing RuntimeHost/tool/approval path remains the only effect authority.** The neural model never calls a tool directly and can never widen capabilities.

Observed planning baselines:
- Zara `master@8a11e9c583a5539c187633bcf87901111e2cc827`.
- `lost-rob0t/zara-plugins@c6d96a01a2c16dd8f41083e281b8d2ef145c18f3`.
- Cross-platform UI extension ABI is already being implemented by #961; consume it rather than creating another UI loader.
- Android already has the `LocalAiService`/local-provider architecture and explicit CPU/GPU/NPU backend policy documented in `wiki/android-local-ai.org`; this feature must not create another assistant/runtime authority.

Related: #647, #623, #152, #961, #655, #622.

---

### ANALYZE

The request has three distinct concerns that should not be collapsed:

1. **Symbolic authority** — Prolog decides whether a learned suggestion is legal/useful.
2. **Adaptive scoring** — a tiny local neural model learns input/tool-sequence correlations.
3. **Companion UX** — desktop Qt and Android Compose render the same semantic status through the in-flight #961 UI contribution ABI.

The neural component is a recommender, not a second planner/runtime. It observes only canonical tool events after Zara has already assigned principal/run/generation identity.

### RESEARCH — 2026-09-17

Current upstream findings:

- Google LiteRT's modern Android `CompiledModel` API provides CPU/GPU/NPU inference selection. LiteRT also documents incremental on-device training using model signatures (`train`, `infer`, `save`, `restore`) through the Interpreter API. NPU availability remains vendor/device dependent, so Zara must capability-probe and report the actual backend rather than claiming TPU/NPU acceleration when it fell back.
- Alibaba MNN currently supports CPU, Vulkan and CUDA backends, plus `MNNTrain`/MNN Express training and model extraction. This makes it a strong desktop candidate for one small model across CPU/Vulkan/CUDA instead of three unrelated implementations.
- #961 already defines native Qt/Compose renderers with semantic UI contributions from Prolog/Python/plugin manifests. The Neurogotchi UI should be a consumer of that ABI.

Primary references:
- https://developers.google.com/edge/litert/android
- https://developers.google.com/edge/litert/conversion/tensorflow/build/ondevice_training
- https://developers.google.com/edge/litert/android/npu/overview
- https://github.com/alibaba/MNN/wiki/engine
- https://github.com/alibaba/MNN/wiki/other

### DESIGN

```text
user input
  + normalized symbolic intent/context
  + recent canonical tool IDs/outcomes
             |
             v
      bounded feature encoder
             |
             v
  Neurogotchi policy model (proposal only)
   next-tool logits + chain/value score
             |
             v
  typed candidate suggestions
             |
             v
        Prolog policy
  allow/reject/rerank/explain
             |
             v
 existing Zara RuntimeHost / ToolRegistry
 approval + principal + generation + execution
             |
             v
 canonical outcome observation
             |
             v
       LearnerActor replay
  bounded online update + checkpoint
```

### Neural model v1

Keep v1 intentionally tiny and auditable:

- hashed/embedded **input feature vector** from local normalized text/intent features;
- learned embeddings for tool IDs and coarse result classes;
- small GRU or equivalent sequence encoder over the last N tool events;
- next-tool classification head;
- optional success/value head used only for ranking;
- fixed maximum vocabulary and explicit `unknown_tool` handling;
- no free-form generated code, arguments, Prolog clauses, shell, or tool names.

The learner trains from bounded episodes containing only approved schema fields. A suggested chain is at most a small configured depth and every step is independently revalidated by Prolog + canonical tool policy.

### Learning / promotion policy

One actor owns replay/training/checkpoint state:

```text
ObserveEpisode
Suggest
TrainTick
EvaluateCandidateCheckpoint
PromoteCheckpoint
RollbackCheckpoint
Snapshot
Shutdown
```

Rules:

- bounded replay buffer, per principal/profile unless explicitly configured otherwise;
- training is opt-in and locally disableable;
- raw credentials/secrets/tool outputs are never training features;
- persisted user text is off by default; prefer symbolic intent IDs, hashed lexical features and redacted/coarse argument classes;
- success/failure/denial/cancel/timeout are distinct outcomes;
- cancelled/stale generations cannot train or score as current state;
- candidate checkpoint must beat or match the current checkpoint on a deterministic holdout/regression corpus before promotion;
- retain last-known-good checkpoint and atomically roll back corruption/regression;
- finite epochs/steps/time/energy budget per train tick;
- no autonomous code/model-topology mutation.

### Prolog API

The plugin/runtime boundary should expose closed predicates conceptually like:

```prolog
neurogotchi_status(Status).
neurogotchi_suggest(Context, RecentTools, Limit, Suggestions).
neurogotchi_candidate_allowed(Context, Candidate, Reason).
neurogotchi_observe(Episode).
neurogotchi_learning_mode(Mode).
neurogotchi_backend(Backend, State).
neurogotchi_mood(Mood, Face, Message).
```

`Suggestions` contains known registered tool IDs + confidence/value metadata only. Prolog may reject, rerank, restrict depth, force deterministic routing, or disable neural suggestions for any capability class.

#### Android

Use Zara's existing local-inference/service ownership; do not start another exported model service.

- **Inference:** LiteRT `.tflite`, preference `NPU -> GPU -> CPU` when supported and explicitly verified. “TPU” is treated as the Android NPU/accelerator path actually exposed by LiteRT/vendor delegates; capability state must name the real backend.
- **Training:** LiteRT on-device training signatures (`train/infer/save/restore`). Probe whether the selected delegate supports the train graph; if not, train locally on CPU and continue accelerated inference. Never report NPU training unless execution evidence proves it.
- use WorkManager/idle/charging/thermal constraints for non-interactive train ticks;
- model/checkpoints remain app-private and version/hash/schema tagged.

#### Desktop Linux

Initial backend candidate: **MNN** with an explicit backend adapter:

```text
auto -> CUDA when verified available
     -> Vulkan when verified available
     -> CPU
```

Build with the relevant MNN CPU/Vulkan/CUDA and Train components. Backend selection must be observable and tested; silent fallback is reported as fallback, not as accelerator success. If research proves a training path is unsupported on a requested accelerator, preserve that backend for inference and use the strongest supported local training backend rather than fabricating support.

The model/checkpoint schema must be portable enough to test equivalent predictions across supported backends.

### Companion / “gotchi” state

Derive the face/mood deterministically from bounded learner state rather than letting a model invent UI text/state. Initial semantic states:

- `booting`
- `curious` — high novelty / low sample count
- `learning` — active bounded train tick
- `confident` — stable high-confidence suggestions
- `confused` — prediction entropy/error spike
- `cooldown` — energy/thermal/budget pause
- `sleeping` — learning disabled/idle
- `degraded` — model/backend/checkpoint unavailable

Expose semantic values such as episodes seen, accepted suggestions, rejected suggestions, holdout score, novelty, current backend, last checkpoint generation and top learned tool sequences. UI maps those to an animated/text face.

### Cross-platform UI

Depend on #961. Do **not** introduce another renderer or plugin UI registry.

Plugin/UI contribution targets:

- drawer/surface entry: `Neurogotchi`;
- compact face/status card in chat or diagnostics;
- full screen: current face/mood, backend, learning state, episode count, confidence/novelty, recent learned chain summaries;
- settings: enable learning, pause, replay size/energy budget presets, backend preference, reset local learned state, export/import checkpoint only through explicit local action;
- debug view: Prolog explanation for accepted/rejected candidate plus model score;
- no raw secret/tool-output display.

Qt and Compose remain native renderers of the same semantic contribution/state.

### Plugin packaging

The public service/plugin implementation belongs in `lost-rob0t/zara-plugins` (proposed name: `zara-neurogotchi`). Zara Core owns only the narrow canonical observation/suggestion/native Android inference seams that cannot live in an external Python plugin.

This does **not** require a separate Android capability APK: there is no new privileged Android permission boundary here. Consume the standard Zara Android UI/local-inference paths and #961 plugin manifest ABI instead of #923/#924's separately installed capability-APK transport.

### ADVERSARIAL REVIEW

Must fail closed for:

- neural suggestion of an unknown/unregistered tool;
- suggestion for a tool current principal cannot use;
- stale tool registry/model/runtime generation;
- prompt/tool output containing secrets or adversarial strings becoming persisted raw training data;
- cross-principal replay/model contamination;
- reward hacking by repeated cheap successful tools;
- feedback loops where model suggestions generate their own positive labels without independent outcome evidence;
- catastrophic forgetting / checkpoint regression;
- corrupt/partial checkpoint;
- enormous/recursive tool chain;
- cancellation during train/save/promote;
- concurrent inference and checkpoint promotion race;
- backend crash/OOM/device loss;
- GPU/NPU backend silently falling back while UI claims accelerated execution;
- thermal/battery runaway on Android;
- UI/plugin state surviving after plugin unload or runtime generation replacement.

### DECISION

Proceed with a **proposal-only local adaptive scorer + Prolog gate + existing canonical executor**, rendered through #961. Use LiteRT on Android and research/validate MNN as the unified desktop CPU/Vulkan/CUDA backend. No wireless offensive behavior is part of this feature.

This adds a durable learned-state authority, new local ML dependency/backend, and canonical tool-observation/suggestion contract, so implementation is **vote-gated by #647**.

---

## Implementation slices after vote

1. **Episode/schema + Prolog policy contract** — canonical redacted `ToolEpisode`, suggestion/result types, principal/generation fencing, deterministic policy tests.
2. **LearnerActor + reference CPU model** — tiny sequence model, replay, holdout promotion/rollback, deterministic synthetic learning test.
3. **Desktop acceleration** — MNN CPU/Vulkan/CUDA build/backend probing, parity/resource tests, exact fallback reporting.
4. **Android learner** — LiteRT train/infer/save/restore, NPU/GPU/CPU capability probe, WorkManager/thermal/lifecycle tests.
5. **Plugin package** — `zara-neurogotchi` service plugin + Prolog module + state/config/docs in `zara-plugins`.
6. **Cross-platform UI** — consume #961 semantic UI ABI in desktop Qt + Android Compose; no second renderer.
7. **Adversarial/e2e gate** — learn synthetic tool patterns, suggest through Prolog, prove denied candidate causes zero effect, checkpoint/restart/backend/lifecycle tests.

Do not start a later slice around an unresolved API/authority decision from an earlier one.

---

## Acceptance

- [ ] architecture vote passes under #647 before production implementation;
- [ ] model can learn at least two deterministic input + tool-sequence patterns from a synthetic episode corpus and improve held-out next-tool accuracy over its untrained baseline;
- [ ] model output is candidate data only; it cannot execute or register a tool;
- [ ] Prolog can reject/rerank candidates and a rejected candidate produces zero effect;
- [ ] all actual effects still cross canonical principal/capability/approval/generation checks;
- [ ] no raw credential/secret/private tool output enters persisted replay/checkpoints/logs;
- [ ] replay/checkpoint/model state is principal/profile isolated and restart-safe;
- [ ] stale/cancelled runtime generations cannot train or emit current suggestions;
- [ ] holdout regression prevents bad checkpoint promotion and rollback restores last-known-good;
- [ ] desktop CPU path is deterministic; Vulkan and CUDA paths are used only when actually verified, with parity/tolerance tests and truthful fallback state;
- [ ] Android LiteRT CPU/GPU/NPU inference path reports actual selected accelerator; on-device training persists/restores weights; unsupported accelerator training falls back explicitly rather than lying;
- [ ] bounded train budgets prevent UI jank/resource runaway; Android thermal/battery constraints are tested;
- [ ] desktop and Android render Neurogotchi through #961-compatible semantic UI contributions with matching state vocabulary;
- [ ] learning can be paused/reset locally and disabled mode performs no training writes;
- [ ] plugin unload/reload and model/checkpoint corruption degrade cleanly;
- [ ] focused tests, full Zara gate, Nix gate, Android gate and `zara-plugins` registry/plugin gates are exact-head green.

---

## VOTE_REQUEST — #647

**Decision requested:** approve a new Prolog-governed local adaptive tool-chain recommender with durable bounded learned state, LiteRT Android backend, MNN desktop CPU/Vulkan/CUDA backend candidate, and UI rendered through existing #961 contribution ABI.

**Affected contracts:** canonical tool-observation/suggestion events, durable local learned-state ownership, local ML dependencies/backends, Android local-inference integration, cross-repo Zara ↔ `zara-plugins` contract.

**Authority position:** neural output is never action authority. Prolog + existing Zara principal/capability/approval/runtime-generation path remains final authority, and the learner cannot create tool names, arguments, code, Prolog clauses, permissions, or capabilities.

**Security/privacy position:** local-only by default; bounded redacted features; per-principal isolation; no raw secrets/tool output in training state; deterministic holdout promotion/rollback; no offensive wireless behavior.
